src/mongo/db/storage/csv_file.cpp

```diff
@@ -0,0 +1,351 @@
+#include "mongo/db/storage/io_error_message.h"
+#include "mongo/logv2/log.h"
+#include "mongo/util/assert_util.h"
+
+#define MONGO_LOGV2_DEFAULT_COMPONENT ::mongo::logv2::LogComponent::kCommand
+namespace mongo {
+using namespace fmt::literals;
+
+CsvFileInput::CsvFileInput(const std::string& fileRelativePath, const std::string& metaFileRelative)
+    : _fileAbsolutePath((externalPipeDir == "" ? kDefaultPipePath : externalPipeDir) +
+                        fileRelativePath),
+      _metadataAbsolutePath((externalPipeDir == "" ? kDefaultPipePath : externalPipeDir) +
+                            metaFileRelative),
+      _ifs() {
+    uassert(200000400,
+            "File path must not include '..' but {} does"_format(_fileAbsolutePath),
+            _fileAbsolutePath.find("..") == std::string::npos);
+    uassert(200000401,
+            "File path must not include '..' but {} does"_format(_metadataAbsolutePath),
+            _metadataAbsolutePath.find("..") == std::string::npos);
+}
+
+CsvFileInput::~CsvFileInput() {
+    _ifs.close();
+}
+
+void CsvFileInput::doOpen() {
+    std::string metadata;
+    std::ifstream ifsMetadata(_metadataAbsolutePath.c_str(), std::ios::in);
+    uassert(ErrorCodes::FileNotOpen,
+            "error = {}: "_format(getErrorMessage("open", _metadataAbsolutePath)),
+            ifsMetadata.is_open());
+    std::getline(ifsMetadata, metadata);
+
+    _metadata = getMetadata(parseLine(metadata));
+    ifsMetadata.close();
+
+    _ifs.open(_fileAbsolutePath.c_str(), std::ios::in);
+
+    uassert(ErrorCodes::FileNotOpen,
+            "error = {}"_format(getErrorMessage("open", _fileAbsolutePath)),
+            _ifs.is_open());
+}
+
+// caller must ensure that buffer size is greater than or equal to the size of the bsonObject to be
+// returned. If not, it would throw exception (Not enough size in buffer)
+int CsvFileInput::doRead(char* data, int size) {
+    auto bsonObj = readBsonObj();
+
+    if (!bsonObj) {
+        return 0;
+    }
+
+    int nRead = bsonObj->objsize();
+    uassert(200000402,
+            "Buff Size {} bytes is too small to contain {} bytes bsonObj"_format(size, nRead),
+            nRead <= size);
+    memcpy(data, bsonObj->objdata(), nRead);
+    return nRead;
+}
+
+void CsvFileInput::doClose() {
+    _ifs.close();
+}
+
+bool CsvFileInput::isOpen() const {
+    return _ifs.is_open();
+}
+
+bool CsvFileInput::isGood() const {
+    return _ifs.good();
+}
+
+bool CsvFileInput::isFailed() const {
+    return _ifs.fail();
+}
+
+bool CsvFileInput::isEof() const {
+    return _ifs.eof();
+}
+
+// assuming that header is returned by parseLine, which means, each of its element is
+// name of the field with typeInfo
+Metadata CsvFileInput::getMetadata(const std::vector<std::string>& header) {
+    constexpr size_t typeNameAbsent = 0;
+    Metadata ret;
+
+    size_t i = 0;
+    for (const auto& field : header) {
+        // Test scenario where / is missing
+        size_t index = field.find('/');
```

Comment:
Can we improve this comment a bit further something like...?
```suggestion
        // Covers both case where a field does not have '/' at all or typename absent after '/'.
        size_t index = field.find('/');
```

Reply:
Done